{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/coretype/tree.py:365: SyntaxWarning: invalid escape sequence '\\-'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/coretype/tree.py:436: SyntaxWarning: invalid escape sequence '\\-'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/coretype/tree.py:1508: SyntaxWarning: invalid escape sequence '\\-'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/coretype/tree.py:365: SyntaxWarning: invalid escape sequence '\\-'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/coretype/tree.py:436: SyntaxWarning: invalid escape sequence '\\-'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/coretype/tree.py:1508: SyntaxWarning: invalid escape sequence '\\-'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:51: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:54: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:55: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:402: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:51: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:54: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:55: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/newick.py:402: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/utils.py:82: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/utils.py:82: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/main.py:64: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/main.py:712: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/main.py:714: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/main.py:64: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/main.py:712: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/main.py:714: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/faces.py:159: SyntaxWarning: invalid escape sequence '\\_'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/treeview/faces.py:159: SyntaxWarning: invalid escape sequence '\\_'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:76: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:108: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:123: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:138: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:76: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:108: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:123: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/parser/phylip.py:138: SyntaxWarning: invalid escape sequence '\\s'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylo/phylotree.py:132: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylo/phylotree.py:188: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylo/phylotree.py:132: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylo/phylotree.py:188: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:74: SyntaxWarning: invalid escape sequence '\\('\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:78: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:82: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:86: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:92: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:217: SyntaxWarning: invalid escape sequence '\\('\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:227: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:240: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:245: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:252: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:258: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:264: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:265: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:74: SyntaxWarning: invalid escape sequence '\\('\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:78: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:82: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:86: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:92: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:217: SyntaxWarning: invalid escape sequence '\\('\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:227: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:240: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:245: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:252: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:258: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:264: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/parser/codemlparser.py:265: SyntaxWarning: invalid escape sequence '\\d'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/evoltree.py:468: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/evoltree.py:471: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/evoltree.py:468: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/evoltree.py:471: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/tools/utils.py:28: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/tools/utils.py:28: SyntaxWarning: invalid escape sequence '\\['\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/model.py:367: SyntaxWarning: invalid escape sequence '\\.'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/model.py:368: SyntaxWarning: invalid escape sequence '\\.'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/model.py:367: SyntaxWarning: invalid escape sequence '\\.'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/evol/model.py:368: SyntaxWarning: invalid escape sequence '\\.'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:75: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:153: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:500: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:756: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:757: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:758: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:75: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:153: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:500: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:756: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:757: SyntaxWarning: invalid escape sequence '\\w'\n",
      "/opt/anaconda3/envs/ete3/lib/python3.12/site-packages/ete3-3.1.3-py3.10.egg/ete3/phylomedb/phylomeDB3.py:758: SyntaxWarning: invalid escape sequence '\\w'\n"
     ]
    }
   ],
   "source": [
    "from ete3 import TreeStyle\n",
    "from ete3 import PhyloTree\n",
    "from ete3 import Tree, faces, TreeStyle\n",
    "from PyQt5       import QtCore\n",
    "from PyQt5.QtWidgets import QGraphicsRectItem\n",
    "from PyQt5.QtGui import QColor, QPen, QBrush\n",
    "\n",
    "from PyQt5.QtWidgets import QGraphicsSimpleTextItem\n",
    "from PyQt5.QtGui import QFont\n",
    "\n",
    "from random import random\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "import re\n",
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Yeast genome order browser alignment stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting the background colors so that only the F is black\n",
    "_aabgcolors = {\n",
    "    'A':\"#FFFFFF\" ,\n",
    "    'R':\"#FFFFFF\" ,\n",
    "    'N':\"#FFFFFF\" ,\n",
    "    'D':\"#FFFFFF\" ,\n",
    "    'C':\"#FFFFFF\" ,\n",
    "    'Q':\"#FFFFFF\" ,\n",
    "    'E':\"#FFFFFF\" ,\n",
    "    'G':\"#FFFFFF\" ,\n",
    "    'H':\"#FFFFFF\" ,\n",
    "    'I':\"#FFFFFF\" ,\n",
    "    'L':\"#FFFFFF\" ,\n",
    "    'K':\"#FFFFFF\" ,\n",
    "    'M':\"#FFFFFF\" ,\n",
    "    'F':\"#000000\" ,\n",
    "    'P':\"#FFFFFF\" ,\n",
    "    'S':\"#FFFFFF\" ,\n",
    "    'T':\"#FFFFFF\" ,\n",
    "    'W':\"#FFFFFF\" ,\n",
    "    'Y':\"#FFFFFF\" ,\n",
    "    'V':\"#FFFFFF\" ,\n",
    "    'B':\"#FFFFFF\" ,\n",
    "    'Z':\"#FFFFFF\" ,\n",
    "    'X':\"#FFFFFF\",\n",
    "    '.':\"#FFFFFF\",\n",
    "    '-':\"#FFFFFF\",\n",
    "    }\n",
    "\n",
    "# Setting the foreground colors so that only F is white\n",
    "_aafgcolors = {\n",
    "    'A':\"#000000\" ,\n",
    "    'R':\"#000000\" ,\n",
    "    'N':\"#000000\" ,\n",
    "    'D':\"#000000\" ,\n",
    "    'C':\"#000000\" ,\n",
    "    'Q':\"#000000\" ,\n",
    "    'E':\"#000000\" ,\n",
    "    'G':\"#000000\" ,\n",
    "    'H':\"#000000\" ,\n",
    "    'I':\"#000000\" ,\n",
    "    'L':\"#000000\" ,\n",
    "    'K':\"#000000\" ,\n",
    "    'M':\"#000000\" ,\n",
    "    'F':\"#FFFFFF\" ,\n",
    "    'P':\"#000000\" ,\n",
    "    'S':\"#000000\" ,\n",
    "    'T':\"#000000\" ,\n",
    "    'W':\"#000000\" ,\n",
    "    'Y':\"#000000\" ,\n",
    "    'V':\"#000000\" ,\n",
    "    'B':\"#000000\" ,\n",
    "    'Z':\"#000000\" ,\n",
    "    'X':\"#000000\",\n",
    "    '.':\"#000000\",\n",
    "    '-':\"#000000\",\n",
    "}\n",
    "\n",
    "# Code from ete package, mofifying to use the colors we set\n",
    "class MySequenceFace(faces.StaticItemFace):\n",
    "    \"\"\" Creates a new molecular sequence face object.\n",
    "\n",
    "\n",
    "    :argument seq:  Sequence string to be drawn\n",
    "    :argument seqtype: Type of sequence: \"nt\" or \"aa\"\n",
    "    :argument fsize:   Font size,  (default=10)\n",
    "\n",
    "    You can set custom colors for amino-acids or nucleotides:\n",
    "\n",
    "    :argument None  codon       : a string that corresponds to the reverse translation of the amino-acid sequence\n",
    "    :argument 11    col_w       : width of the column (if col_w is lower than font size, letter wont be displayed)\n",
    "    :argument None  fg_colors   : dictionary of colors for foreground, with as keys each possible character in sequences, and as value the colors\n",
    "    :argument None  bg_colors   : dictionary of colors for background, with as keys each possible character in sequences, and as value the colors\n",
    "    :argument 3     alt_col_w   : works together with special_col option, defines the width of given columns\n",
    "    :argument None  special_col : list of lists containing the bounds of columns to be displayed with alt_col_w as width\n",
    "    :argument False interactive : more info can be displayed when mouse over sequence\n",
    "\n",
    "    \"\"\"\n",
    "    def __init__(self, seq, seqtype=\"aa\", fsize=10,\n",
    "                 fg_colors=None, bg_colors=None,\n",
    "                 codon=None, col_w=11, alt_col_w=3,\n",
    "                 special_col=None, interactive=False):\n",
    "        self.seq         = seq\n",
    "        self.codon       = codon\n",
    "        self.fsize       = fsize\n",
    "        self.style       = seqtype\n",
    "        self.col_w       = float(col_w)\n",
    "        self.alt_col_w   = float(alt_col_w)\n",
    "        self.special_col = special_col if special_col else []\n",
    "        self.width       = 0 # will store the width of the whole sequence\n",
    "        self.interact    = interactive\n",
    "\n",
    "        if self.style == \"aa\":\n",
    "            if not fg_colors:\n",
    "                fg_colors = _aafgcolors\n",
    "            if not bg_colors:\n",
    "                bg_colors = _aabgcolors\n",
    "        else:\n",
    "            if not fg_colors:\n",
    "                fg_colors = _ntfgcolors\n",
    "            if not bg_colors:\n",
    "                bg_colors = _ntbgcolors\n",
    "\n",
    "        self.fg_col = self.__init_col(fg_colors)\n",
    "        self.bg_col = self.__init_col(bg_colors)\n",
    "\n",
    "        # for future?\n",
    "        self.row_h       = 13.0\n",
    "\n",
    "        super(MySequenceFace,\n",
    "              self).__init__(QGraphicsRectItem(0, 0, self.width, self.row_h))\n",
    "\n",
    "    def __init_col(self, color_dic):\n",
    "        \"\"\"to speed up the drawing of colored rectangles and characters\"\"\"\n",
    "        new_color_dic = {}\n",
    "        for car in color_dic:\n",
    "            new_color_dic[car] = QBrush(QColor(color_dic[car]))\n",
    "        return new_color_dic\n",
    "\n",
    "    def update_items(self):\n",
    "        #self.item = QGraphicsRectItem(0,0,self._total_w, self.row_h)\n",
    "        seq_width = 0\n",
    "        nopen = QPen(QtCore.Qt.NoPen)\n",
    "        font = QFont(\"Courier\", self.fsize)\n",
    "        rect_cls = QGraphicsRectItem\n",
    "        for i, letter in enumerate(self.seq):\n",
    "            width = self.col_w\n",
    "            for m in self.special_col:\n",
    "                if m[0] < i <= m[1]:\n",
    "                    width = self.alt_col_w\n",
    "                    break\n",
    "            #load interactive item if called correspondingly\n",
    "            rectItem = rect_cls(0, 0, width, self.row_h, parent=self.item)\n",
    "            rectItem.setX(seq_width) # to give correct X to children item\n",
    "            rectItem.setBrush(self.bg_col[letter])\n",
    "            rectItem.setPen(nopen)\n",
    "            if self.interact:\n",
    "                if self.codon:\n",
    "                    rectItem.codon = '%s, %d: %s' % (self.seq[i], i,\n",
    "                                                     self.codon[i*3:i*3+3])\n",
    "                else:\n",
    "                    rectItem.codon = '%s, %d' % (self.seq[i], i)\n",
    "            # write letter if enough space\n",
    "            if width >= self.fsize:\n",
    "                text = QGraphicsSimpleTextItem(letter, parent=rectItem)\n",
    "                text.setFont(font)\n",
    "                text.setBrush(self.fg_col[letter])\n",
    "                # Center text according to rectItem size\n",
    "                tw = text.boundingRect().width()\n",
    "                th = text.boundingRect().height()\n",
    "                text.setPos((width - tw)/2, (self.row_h - th)/2)\n",
    "            seq_width += width\n",
    "        self.width = seq_width"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a subset of the alignment\n",
    "def subset_alignment(alignment_file, start, end):\n",
    "    genes = []\n",
    "    seqs = []\n",
    "    seq = \"\"\n",
    "\n",
    "    # Read in alignment sequences and save to list\n",
    "    with open(alignment_file, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "    for line in lines:\n",
    "        if line.startswith(\">\"):\n",
    "            genes.append(line.strip())\n",
    "            if len(seq) > 0:\n",
    "                seqs.append(seq)\n",
    "                seq = \"\"\n",
    "        else:\n",
    "            seq += line.strip()\n",
    "    seqs.append(seq)\n",
    "    seq = \"\"\n",
    "\n",
    "    # Write it to a string in fasta format\n",
    "    fasta_txt = \"\"\n",
    "    for gene, seq in zip(genes, seqs):\n",
    "        fasta_txt += gene + \"\\n\"\n",
    "        fasta_txt += str(seq[start:end]) + \"\\n\"\n",
    "    \n",
    "    return fasta_txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warnning: [3] terminal nodes could not be found in the alignment.\n"
     ]
    }
   ],
   "source": [
    "# Layout function (adds things to node) to add the alignment\n",
    "def test_layout_phylo_aa(node):\n",
    "    '''\n",
    "    layout for CodemlTree\n",
    "    '''\n",
    "    if hasattr(node, \"collapsed\"):\n",
    "        if node.collapsed == 1:\n",
    "            node.img_style[\"draw_descendants\"]= False\n",
    "    if node.is_leaf:\n",
    "        if hasattr (node, \"sequence\"):\n",
    "            seqface =  MySequenceFace(node.sequence, \"aa\",\n",
    "                                      fsize=10,\n",
    "                                      col_w=11, interactive=False)\n",
    "            faces.add_face_to_node(seqface, node, 1, aligned=True)\n",
    "\n",
    "\n",
    "alignment_file = \"gcn4_paper_stuff/gain_loss_Fs/YGOB_preComputed_MSA.fasta\"\n",
    "\n",
    "\n",
    "fasta_txt = subset_alignment(alignment_file, 149, 223)\n",
    "\n",
    "tree = PhyloTree(\"gcn4_paper_stuff/gain_loss_Fs/ygob_gene_tree.tree\")\n",
    "tree.link_to_alignment(alignment=fasta_txt, alg_format=\"fasta\")\n",
    "\n",
    "# Define your start and end positions\n",
    "start_position = 30\n",
    "end_position = 55\n",
    "\n",
    "# Create the TextFace objects for the start and end positions\n",
    "start_annotation = faces.TextFace(f\"{start_position}\", fsize=10, fgcolor=\"black\")\n",
    "end_annotation = faces.TextFace(f\"{end_position}\", fsize=10, fgcolor=\"black\")\n",
    "\n",
    "leafs = tree.get_leaf_names()\n",
    "clade_leaves = [leaf for leaf in leafs if leaf in fasta_txt]\n",
    "tree.prune(clade_leaves)\n",
    "\n",
    "first_leaf = tree.get_leaves()[0]\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.layout_fn = test_layout_phylo_aa\n",
    "# Add the start and end annotations to the first leaf\n",
    "first_leaf.add_face(start_annotation, column=0, position=\"aligned\")\n",
    "\n",
    "tree.show(tree_style=ts)\n",
    "# tree.render(\"gcn4_paper_stuff/gain_loss_Fs/ygob_MSA_gene_tree.pdf\", tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warnning: [3] terminal nodes could not be found in the alignment.\n"
     ]
    }
   ],
   "source": [
    "def test_layout_phylo_aa(node):\n",
    "    '''\n",
    "    layout for CodemlTree\n",
    "    '''\n",
    "    if hasattr(node, \"collapsed\"):\n",
    "        if node.collapsed == 1:\n",
    "            node.img_style[\"draw_descendants\"]= False\n",
    "    if node.is_leaf:\n",
    "        if hasattr (node, \"sequence\"):\n",
    "            seqface =  MySequenceFace(node.sequence, \"aa\",\n",
    "                                      fsize=10,\n",
    "                                      col_w=11, interactive=False)\n",
    "            faces.add_face_to_node(seqface, node, 1, aligned=True)\n",
    "\n",
    "\n",
    "alignment_file = \"gcn4_paper_stuff/gain_loss_Fs/ygob_MSA_species_names.fasta\"\n",
    "\n",
    "\n",
    "fasta_txt = subset_alignment(alignment_file, 149, 223)\n",
    "\n",
    "tree = PhyloTree(\"gcn4_paper_stuff/gain_loss_Fs/ygob_species_tree.tree\")\n",
    "tree.link_to_alignment(alignment=fasta_txt, alg_format=\"fasta\")\n",
    "\n",
    "# Define your start and end positions\n",
    "start_position = 30\n",
    "end_position = 55\n",
    "\n",
    "# Create the TextFace objects for the start and end positions\n",
    "start_annotation = faces.TextFace(f\"{start_position}\", fsize=10, fgcolor=\"black\")\n",
    "end_annotation = faces.TextFace(f\"{end_position}\", fsize=10, fgcolor=\"black\")\n",
    "\n",
    "leafs = tree.get_leaf_names()\n",
    "clade_leaves = [leaf for leaf in leafs if leaf in fasta_txt]\n",
    "tree.prune(clade_leaves)\n",
    "\n",
    "first_leaf = tree.get_leaves()[0]\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.layout_fn = test_layout_phylo_aa\n",
    "# Add the start and end annotations to the first leaf\n",
    "first_leaf.add_face(start_annotation, column=0, position=\"aligned\")\n",
    "\n",
    "tree.show(tree_style=ts)\n",
    "# tree.render(\"gcn4_paper_stuff/gain_loss_Fs/ygob_MSA_species_tree.pdf\", tree_style=ts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ete3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
