{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c96e102",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import localcider\n",
    "# from localcider.sequenceParameters import SequenceParameters\n",
    "# import protfasta\n",
    "# import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import re\n",
    "# import scipy\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "# from collections import defaultdict \n",
    "# from Bio.Seq import Seq\n",
    "# import Bio\n",
    "# import matplotlib.patches as mpatches\n",
    "from matplotlib.patches import Patch\n",
    "from scipy.signal import find_peaks\n",
    "\n",
    "import seaborn as sns\n",
    "lightblue , darkblue, lightgreen, darkgreen, pink =  '#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99'\n",
    "# import editdistance\n",
    "\n",
    "# import YeastAnalysisfunctions as YAF\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "plt.rcParams[\"font.family\"] = \"Helvetica\"\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "# alignment distance vs activity difference \n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85aef90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "masterDF = pd.read_csv('SC_Glucose_Ratio_MasterDF_20240930_withSeqFeatures.csv')\n",
    "maxActivity = max(masterDF.Activity_SCglucose)\n",
    "minActivity = min(masterDF[masterDF.Recovered].Activity_SCglucose)\n",
    "\n",
    "Activity = 'Activity_SCglucose'\n",
    "print('%i of %i Designed ADs recovered, %.1f percent'%(sum(masterDF['Recovered']),len(masterDF),100.0*sum(masterDF['Recovered'])/len(masterDF)))\n",
    "\n",
    "\n",
    "# ActivityThresh = 100000\n",
    "# ActivityThresh=45373 #new thresholds from Aaron's model June 2024\n",
    "# HighActivityThresh=45373*2\n",
    "ActivityThresh=  94031 # 80th percentile 20241001\n",
    "\n",
    "\n",
    "#define WT\n",
    "ScerWTactivity = np.mean(masterDF[masterDF.ADseq=='STDSTPMFEYENLEDNSKEWTSLFDNDIPVTTDDVSLADK'][Activity])\n",
    "ScerWT_DF = masterDF[masterDF.ADseq=='STDSTPMFEYENLEDNSKEWTSLFDNDIPVTTDDVSLADK']\n",
    "\n",
    "Orthologs_DF = pd.read_csv('../data/OrthologTilingDFwActivities_20240930.csv', index_col=0)\n",
    "\n",
    "FullOrthoDF =  pd.read_pickle('../data/FullOrthologDF_20240930')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4577bf9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Orthologs_DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f5b3b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_peak_ADs(tempDF):\n",
    "    peaks, _ = find_peaks(tempDF[\"SmoothedActivitesLoess\"], prominence=25000, height=ActivityThresh)\n",
    "    if tempDF[\"SmoothedActivitesLoess\"][0] > ActivityThresh + 25000:\n",
    "        peaks = np.append(peaks, 0)\n",
    "\n",
    "    return peaks, len(peaks)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdea57e5",
   "metadata": {},
   "source": [
    "# make a pretty WT Gcn4 plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd340e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# other features\n",
    "Gcn4Features = {\"uORF1\":[-360./3,-349/3,'gray'] ,\n",
    "'uORF2':[-292./3,-284/3,'gray'],\n",
    "'uORF3':[-175./3,-164/3,'gray'],\n",
    "'uORF4':[-150./3,-139/3,'gray'],\n",
    "                'DBD':[225,279,'darkblue'],\n",
    "                'N terminal AD':[0,100,'lightblue'],\n",
    "                'Central acidic AD':[101,140,'orange']\n",
    "               }\n",
    "# 'CDS':[1..846]# (includes STOP)\n",
    "tracecolor ='pink'\n",
    "x = 'NativeLocation'# \n",
    "y = 'Activity'\n",
    "halfwindow =20 \n",
    "fig = plt.figure(figsize=[12,6])\n",
    "ax = fig.add_subplot(1,1,1)\n",
    "# Define the Scer WT for comparison\n",
    "ScertempDF = pd.read_csv('Scer_WT_Data.csv')\n",
    "\n",
    "ax.plot(ScertempDF[x],ScertempDF[y],c=tracecolor,linewidth=2,marker='.',label='AD activity')\n",
    "# ax.scatter(ScertempDF[x],ScertempDF[y],c=tracecolor,linewidth=2,marker='o',label='AD activity')\n",
    "ax.errorbar(ScertempDF[x],ScertempDF[y],yerr=ScertempDF['Stdev_BYS4_BioReps'],c='pink',mfc=tracecolor,linewidth=1,marker='.',mew=0,label='AD activity')\n",
    "\n",
    "\n",
    "for i, entry in ScertempDF.iterrows():\n",
    "    ax.plot([entry[x]+1-halfwindow,entry[x]+1+halfwindow],[entry[y],entry[y]],linewidth=2,c=tracecolor,alpha=.5)\n",
    "ax.plot([-130,281],[ActivityThresh,ActivityThresh],c='gray')\n",
    "tempDF = FullOrthoDF[FullOrthoDF.SpeciesName=='Saccharomycotina_sp|P03069|GCN4_YEAST']\n",
    "# ax.plot(tempDF.SmoothedActivites.values[0],label='Average smoothening')\n",
    "ax.plot(tempDF.SmoothedActivitesLoess.values[0],label='Loess Smoothening')\n",
    "# ax.plot(tempDF.SmoothedActivites)\n",
    "\n",
    "for i in Gcn4Features.keys():\n",
    "    x1,x2,color = Gcn4Features[i]\n",
    "    ax.plot([x1,x2],[1,1],c=color,linewidth=10,alpha=.5,label=i)\n",
    "plt.xlabel('Position (AA)'),plt.ylabel('Activity')\n",
    "plt.legend(loc=2)\n",
    "# plt.savefig('../Results/AnnotatedGcn4_smoothing.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73829eda",
   "metadata": {},
   "outputs": [],
   "source": [
    "domain_df = pd.read_table(\"../data/FullLengthUniqueGcn4Orthologs_domains.tsv\", header=None)\n",
    "columns = [\n",
    "    \"name\",\n",
    "    \"sequence_md5_digest\",\n",
    "    \"length\",\n",
    "    \"analysis\", # Pfam, PRINTS, Gene3D\n",
    "    \"signature_acession\",\n",
    "    \"signature_description\",\n",
    "    \"start\",\n",
    "    \"end\",\n",
    "    \"score\",\n",
    "    \"status\", # T if true?\n",
    "    \"date\",\n",
    "    \"interpro_annot\",\n",
    "    \"interpro_description\",\n",
    "    \"GO_annotations\",\n",
    "    \"Pathway_annotations\"\n",
    "]\n",
    "\n",
    "domain_df.columns = columns\n",
    "cerevisiae_gene = 'Saccharomycotina_sp|P03069|GCN4_YEAST'\n",
    "cerevisiae_domain_df = domain_df[[f\"{cerevisiae_gene}\" in name for name in domain_df[\"name\"]]]\n",
    "SUPERFAMILY = cerevisiae_domain_df[cerevisiae_domain_df[\"analysis\"] == \"SUPERFAMILY\"][\"signature_description\"].values\n",
    "filtered_domain_df = domain_df[[(s in SUPERFAMILY) for s in domain_df[\"signature_description\"]]]\n",
    "domain_coordinates = filtered_domain_df[['name', 'start', 'end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c39a077",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_tile_ADs(tile_tempDF):\n",
    "    ADs = []\n",
    "    AD_end = 0\n",
    "    AD_start = 0\n",
    "\n",
    "    for i in tile_tempDF.index: \n",
    "        if tile_tempDF.loc[i, \"Activity\"] > ActivityThresh:\n",
    "            # If start of new active tile is after the end of last active tile\n",
    "            if tile_tempDF.loc[i, \"NativeLocation\"] - 19 > AD_end:\n",
    "                # Save old AD and start a new AD\n",
    "                if AD_end > 0: \n",
    "                    ADs.append((AD_start, AD_end))\n",
    "                # Make new AD\n",
    "                AD_end = tile_tempDF.loc[i, \"NativeLocation\"] + 20\n",
    "                AD_start = tile_tempDF.loc[i, \"NativeLocation\"] - 19\n",
    "\n",
    "            # Extend current AD\n",
    "            else: \n",
    "                AD_end = tile_tempDF.loc[i, \"NativeLocation\"] + 20\n",
    "\n",
    "    # Add the final AD to the list\n",
    "    if AD_end > 0: \n",
    "        ADs.append((AD_start, AD_end))\n",
    "    \n",
    "    return ADs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66d1c3f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Version for all species\n",
    "def make_activity_trace(name, species, ax):\n",
    "    tracecolor ='pink'\n",
    "    x = 'NativeLocation'# \n",
    "    y = 'Activity'\n",
    "    halfwindow =20 \n",
    "\n",
    "    # Plot tiles first\n",
    "    tile_tempDF = Orthologs_DF[Orthologs_DF[\"SpeciesNames\"] == name].sort_values(\"NativeLocation\").dropna(subset=\"Activity\")\n",
    "\n",
    "    ax.plot(tile_tempDF[x],tile_tempDF[y],c=tracecolor,linewidth=2,marker='.',label='AD activity')\n",
    "    ax.errorbar(tile_tempDF[x],tile_tempDF[y],yerr=tile_tempDF['Stdev_BYS4_BioReps'],c='pink',mfc=tracecolor,linewidth=1,marker='.',mew=0)\n",
    "\n",
    "\n",
    "    # Plotting tile lengths\n",
    "    for i, entry in tile_tempDF.iterrows():\n",
    "        ax.plot([entry[x]+1-halfwindow,entry[x]+1+halfwindow],[entry[y],entry[y]],linewidth=2,c=tracecolor,alpha=.5)\n",
    "\n",
    "    tempDF = FullOrthoDF[FullOrthoDF.SpeciesName==name]\n",
    "    \n",
    "    # Plotting threshold\n",
    "    ax.plot([0,len(tempDF.Seq.values[0])],[ActivityThresh,ActivityThresh],c='gray')\n",
    "\n",
    "    # Plotting the blue line\n",
    "    ax.plot(tempDF.SmoothedActivitesLoess.values[0],label='Loess Smoothening')\n",
    "\n",
    "    # WxxLF location\n",
    "    ax.plot([tempDF.WxxLF_loc,tempDF.WxxLF_loc+4],[1,1],c=\"orange\",linewidth=10,alpha=.5)\n",
    "\n",
    "    # DBD location\n",
    "    DBD_tempDF = domain_coordinates[domain_coordinates[\"name\"] == name]\n",
    "    ax.plot([DBD_tempDF.start,DBD_tempDF.end],[1,1],c=\"darkblue\",linewidth=10,alpha=.5)\n",
    "\n",
    "    # AD locations\n",
    "    AD_ls = get_overlapping_tile_ADs(tile_tempDF)\n",
    "    for start, end in AD_ls:\n",
    "        ax.plot([start,end],[-10000,-10000],c=\"lightblue\",linewidth=10,alpha=.5)\n",
    "\n",
    "    peaks, peak_len = count_peak_ADs(tempDF.iloc[0])\n",
    "\n",
    "    # Add label with number of ADs\n",
    "    if len(AD_ls) != 1:\n",
    "        ax.text(\n",
    "            0.98, 0.7, f'{len(AD_ls)} ADs (overlapping tiles)',\n",
    "            transform=ax.transAxes,\n",
    "            ha='right', va='top',  # horizontal and vertical alignment\n",
    "            fontsize=12,\n",
    "            bbox=dict(facecolor='white', edgecolor='none', alpha=0.7)  # optional background\n",
    "        )\n",
    "    else: \n",
    "        ax.text(\n",
    "            0.98, 0.7, f'{len(AD_ls)} AD (overlapping tiles)',\n",
    "            transform=ax.transAxes,\n",
    "            ha='right', va='top',  # horizontal and vertical alignment\n",
    "            fontsize=12,\n",
    "            bbox=dict(facecolor='white', edgecolor='none', alpha=0.7)  # optional background\n",
    "        )\n",
    "\n",
    "    if peak_len != 1:\n",
    "        ax.text(\n",
    "            0.98, 0.6, f'{peak_len} AD peaks',\n",
    "            transform=ax.transAxes,\n",
    "            ha='right', va='top',  # horizontal and vertical alignment\n",
    "            fontsize=12,\n",
    "            bbox=dict(facecolor='white', edgecolor='none', alpha=0.7)  # optional background\n",
    "        )\n",
    "    else: \n",
    "        ax.text(\n",
    "            0.98, 0.6, f'{peak_len} AD peak',\n",
    "            transform=ax.transAxes,\n",
    "            ha='right', va='top',  # horizontal and vertical alignment\n",
    "            fontsize=12,\n",
    "            bbox=dict(facecolor='white', edgecolor='none', alpha=0.7)  # optional background\n",
    "        )\n",
    "\n",
    "\n",
    "    plt.xlabel('Position (AA)'),plt.ylabel('Activity')\n",
    "\n",
    "    # Manually adding AD to the legend\n",
    "    \n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    custom_patch = Patch(color='lightblue', label='AD')\n",
    "    handles.append(custom_patch)\n",
    "    labels.append('AD')\n",
    "\n",
    "    custom_patch = Patch(color='orange', label='WxxLF')\n",
    "    handles.append(custom_patch)\n",
    "    labels.append('WxxLF')\n",
    "\n",
    "    custom_patch = Patch(color='darkblue', label='DBD')\n",
    "    handles.append(custom_patch)\n",
    "    labels.append('DBD')\n",
    "\n",
    "    ax.legend(handles, labels, loc=1)\n",
    "    ax.set_title(f\"{species}: {name}\")\n",
    "    # plt.savefig(f'figures/{name}.pdf', format=\"pdf\", bbox_inches='tight')\n",
    "    # plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4244ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "name_key = pd.read_csv(\"../data/phylogenetic_info/species_mapping.csv\")\n",
    "name_key = name_key.rename({'SpeciesName' : 'Species'}, axis=1)\n",
    "name_key = name_key.merge(FullOrthoDF, left_on='HomologName', right_on='SpeciesName')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "100fa870",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_peak_ADs(tempDF):\n",
    "    # Prominence was 38000\n",
    "    peaks, _ = find_peaks(tempDF[\"SmoothedActivitesLoess\"], prominence=25000, height=ActivityThresh, distance=15)\n",
    "    # print(peaks)\n",
    "    if tempDF[\"SmoothedActivitesLoess\"][0] > ActivityThresh + 10000:\n",
    "        # Making sure we haven't already called this peak\n",
    "        if len(peaks) == 0 or (tempDF[\"SmoothedActivitesLoess\"][:peaks[0]] <  tempDF[\"SmoothedActivitesLoess\"][0] - 7000).any():\n",
    "            peaks = np.append(peaks, 0)\n",
    "\n",
    "    return peaks, len(peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59625ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1,1)\n",
    "\n",
    "make_activity_trace(\"Anid_AN3675\", \"Plectosphaerella cucumerina\", axs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07f95eda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try to make one big figure\n",
    "rows = 63\n",
    "columns = 8 \n",
    "fig, axs = plt.subplots(rows, columns, figsize=[12*columns,6*rows])\n",
    "\n",
    "i = 0\n",
    "j = 0\n",
    "\n",
    "for species in name_key[\"Species\"].unique():\n",
    "    tempSpeciesDF = name_key[name_key[\"Species\"] == species].sort_values(\"Length\", ascending=False)\n",
    "\n",
    "    for name in tempSpeciesDF[\"HomologName\"]:\n",
    "        make_activity_trace(name, species, axs[i, j])\n",
    "        if j + 1 == columns:\n",
    "            j = 0\n",
    "            i+=1\n",
    "        else:\n",
    "            j += 1\n",
    "\n",
    "while (i < rows) & (j < columns):\n",
    "    axs[i, j].remove()\n",
    "    if j + 1 == columns:\n",
    "        j = 0\n",
    "        i+=1\n",
    "    else:\n",
    "        j += 1\n",
    "\n",
    "\n",
    "plt.savefig(f'figures/all_traces_w_peak_count.pdf', format=\"pdf\", bbox_inches='tight')\n",
    "plt.savefig(f'figures/all_traces_w_peak_count.png', format=\"png\", bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfbab92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try to make one big figure\n",
    "rows = 14\n",
    "columns = 5 \n",
    "fig, axs = plt.subplots(rows, columns, figsize=[12*columns,6*rows])\n",
    "\n",
    "i = 0\n",
    "j = 0\n",
    "iteration = 1\n",
    "\n",
    "for species in name_key[\"Species\"].unique():\n",
    "    tempSpeciesDF = name_key[name_key[\"Species\"] == species].sort_values(\"Length\", ascending=False)\n",
    "\n",
    "    for name in tempSpeciesDF[\"HomologName\"]:\n",
    "        # print(i, j)\n",
    "        make_activity_trace(name, species, axs[i, j])\n",
    "        if j + 1 == columns:\n",
    "            j = 0\n",
    "            i+=1\n",
    "        else:\n",
    "            j += 1\n",
    "\n",
    "        if (i + 1 > rows): \n",
    "            # print(i, j)\n",
    "            plt.savefig(f'figures/all_traces_w_peak_count_v{iteration}.pdf', format=\"pdf\", bbox_inches='tight')\n",
    "            plt.savefig(f'figures/all_traces_w_peak_count_v{iteration}.png', format=\"png\", bbox_inches='tight')\n",
    "            iteration += 1\n",
    "            i = 0\n",
    "            j = 0\n",
    "            fig, axs = plt.subplots(rows, columns, figsize=[12*columns,6*rows])\n",
    "\n",
    "while (i < rows) & (j < columns):\n",
    "    axs[i, j].remove()\n",
    "    if j + 1 == columns:\n",
    "        j = 0\n",
    "        i+=1\n",
    "    else:\n",
    "        j += 1\n",
    "\n",
    "\n",
    "plt.savefig(f'figures/all_traces_w_peak_count_v{iteration}.pdf', format=\"pdf\", bbox_inches='tight')\n",
    "plt.savefig(f'figures/all_traces_w_peak_count_v{iteration}.png', format=\"png\", bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bf62469",
   "metadata": {},
   "outputs": [],
   "source": [
    "# How max calculated number of ADs \n",
    "HighActivityThresh = 150000\n",
    "plothighactive = False \n",
    "\n",
    "tempDict ={}\n",
    "# capture stats about TFs\n",
    "NumActive,NumHighActive, lengths = [],[],[]\n",
    "NumADs,NumHighADs =[],[]\n",
    "NumTFswOneAD,NumTFswOneAD_high,NumTFswZeroAD_high =0,0,0\n",
    "Collect_list_of_ADs =[]\n",
    "\n",
    "FullTFSeqDict = dict(zip(FullOrthoDF.SpeciesName,FullOrthoDF.Seq))\n",
    "\n",
    "#capture ADs that don't overlap the WxxLF motif\n",
    "plt.figure(figsize=(10,2))\n",
    "ADseqs, SpeciesNames = [],[]\n",
    "TileLength = 40\n",
    "genenames, starts,ends, seqs,regionlengths,regionlengthsHigh,RegionType = [],[],[],[],[],[],[]\n",
    "SingleADoverlapsWxxLF,SingleADoverlapsWxxLF_Strong =0,0\n",
    "LengthDict = dict(zip(FullOrthoDF.SpeciesName,FullOrthoDF.Length))\n",
    "\n",
    "names = []\n",
    "# Going through the tiles\n",
    "for ii, name in enumerate( set(Orthologs_DF.SpeciesNames)):\n",
    "    names.append(name)\n",
    "    indx = Orthologs_DF.SpeciesNames==name\n",
    "    \n",
    "    tempDF = Orthologs_DF[indx] # DF for all tiles for each ortholog\n",
    "\n",
    "    lengths.append(LengthDict[name])\n",
    "    #combine overlapping tiles to estimate number of ADs. \n",
    "    tempDF2 = tempDF[tempDF.Activity>ActivityThresh]#active tiles\n",
    "    tempDF2 = tempDF2.sort_values(by='NativeLocation',ascending=True)\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    #check if there are no ADs\n",
    "    if len(tempDF2)==0:\n",
    "        NumADs.append(0)\n",
    "        print(name+' has no active ADs')\n",
    "    else:\n",
    "        tempNumADs =1\n",
    "        previousstart = (tempDF2.NativeLocation[0]-19)\n",
    "        previousend   = (tempDF2.NativeLocation[0]+20)\n",
    "        for i, entry in tempDF2.iterrows():\n",
    "            if (entry.NativeLocation-19) < (previousend):\n",
    "                previousend = entry.NativeLocation+20\n",
    "            else:\n",
    "                starts.append(previousstart)\n",
    "                ends.append(previousend)\n",
    "                tempNumADs +=1\n",
    "                \n",
    "                #restart\n",
    "                previousstart = entry.NativeLocation-19\n",
    "                previousend   = entry.NativeLocation+20\n",
    "            # get seq of the AD \n",
    "            fullseq = FullTFSeqDict[name]\n",
    "\n",
    "        NumADs.append(tempNumADs)      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9dffa55",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_tile_ADs(tile_tempDF):\n",
    "    ADs = []\n",
    "    AD_end = 0\n",
    "    AD_start = 0\n",
    "\n",
    "    for i in tile_tempDF.index: \n",
    "        if tile_tempDF.loc[i, \"Activity\"] > ActivityThresh:\n",
    "            # If start of new active tile is after the end of last active tile\n",
    "            if tile_tempDF.loc[i, \"NativeLocation\"] - 19 > AD_end:\n",
    "                # Save old AD and start a new AD\n",
    "                if AD_end > 0: \n",
    "                    ADs.append((AD_start, AD_end))\n",
    "                # Make new AD\n",
    "                AD_end = tile_tempDF.loc[i, \"NativeLocation\"] + 20\n",
    "                AD_start = tile_tempDF.loc[i, \"NativeLocation\"] - 19\n",
    "\n",
    "            # Extend current AD\n",
    "            else: \n",
    "                AD_end = tile_tempDF.loc[i, \"NativeLocation\"] + 20\n",
    "\n",
    "    # Add the final AD to the list\n",
    "    if AD_end > 0: \n",
    "        ADs.append((AD_start, AD_end))\n",
    "    \n",
    "    return ADs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e689862d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Slight difference but I think we want to use > here because we want to be inclusive of the start? \n",
    "for name, AD_count in zip(names, NumADs):\n",
    "    df = Orthologs_DF[Orthologs_DF[\"SpeciesNames\"] == name].sort_values(\"NativeLocation\").dropna(subset=\"Activity\")\n",
    "    my_count = len(get_overlapping_tile_ADs(df))\n",
    "    if AD_count != my_count:\n",
    "        print(name, AD_count, my_count)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basic_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
